### DISCUSSION TITLE
setLoop

### DISCUSSION THREAD

**Discussion by jialarcon** [pschatzmann/arduino-audio-tools#969] [Q&A] [answered]
Hi, I have tried the Starwars example on my system and it works perfectly but later I tried to replay the file with the setLoop(true) and in did not work, I get a compiler error. I have tried to search for any discussions or examples regarding this subject and did not find any.

music.setLoop(true) just afterwards the MemoryStream music constructor fails to compile with the following error:

'music' does not name a type

Any clues ?

Kind Regards

Jose

**Comment by pschatzmann** [comment, marked as answer]
With the provided information I am not able to figure out what sketch you are referring to nor am I able to understand what you did. If you want me to help, you need to be much more specific and clear.

Here is an [example](https://github.com/pschatzmann/arduino-audio-tools/blob/c5e38e96ab26788f1c1623eb3e25fc12fe87ae67/examples/examples-audiokit/streams-memory_pcm-mixer-audiokit/streams-memory_pcm-mixer-audiokit.ino#L26) that uses the setLoop() method.

I suggest that you read some more C++ tutorials an try to learn to understand the compiler messages...

**Comment by jialarcon** [comment]
Thak you for your quick response, I have noticed in the example you provided that the **setLoop(true)** statement is in the **setup()** section.

In my code it was just after the constructor of the **MemoryStream** object which is above the setup() section

After changing that statement to the **setup()** section it compiled succesfully.

I am not an expert in C++ but I thought I had a working knowledge... I still don't understand in this case why this error happened.

**Reply by joba-1** [reply]
A pretty basic concept slipped through: unlike e.g. python programs which start execution at the beginning of a file, C or C++ programs start with the function main(), which can call other functions which call other functions. This means a function call outside of other functions makes no sense and the compiler complains.

**Reply by jialarcon** [reply]
I am referring to the Arduino IDE structure in which you have as you know three blocks: 

The first one with "includes" and global variables and constant definitions.

The second one the setup() and the third one the loop()

In the  first block you can initialize variables, construct objects and I thought that you could also set class variables, execute class methods and so on, but I have checked that this type of code is not allowed in that section and that is the reason the compiling was failing. Sometimes compiler errors are cryptic or misleading...

It has been a few months since I had used the Arduino ide and I must have forgotten about this detail. 

PS I have never used Python or other similar program.

Anyway thank you for your comments, having a look at the example showed me the way.

Kind Regards
